nnvm/src/compiler/graph_fuse.cc

```diff
@@ -161,6 +161,44 @@ nnvm::Graph GraphFusePartition(nnvm::Graph g) {
       }
     }
   }
+
+  if (opt_level >= 1) {
+    std::vector<std::vector<uint32_t>> children_group_ids(idx.num_nodes());
+    std::vector<std::vector<uint32_t>> node_ids_per_group(idx.num_nodes());
+    for (uint32_t nid = idx.num_nodes() - 1; nid != 0; --nid) {
+      const auto& inode = idx[nid];
+      if (inode.source->is_variable()) continue;
+      CHECK_NE(group_vec[nid], -1);
+      node_ids_per_group[group_vec[nid]].push_back(nid);
+      if (inode.inputs.size() != 1) continue;
+      const auto& parent_nid = inode.inputs[0].node_id;
+      // if parent node has more than one child, record each child's group id.
+      if (ref_count[parent_nid] > 1) children_group_ids[parent_nid].push_back(group_vec[nid]);
+    }
+    std::vector<int> new_group_id(idx.num_nodes(), -1);
+    for (uint32_t nid = idx.num_nodes() - 1; nid != 0; --nid) {
+      if (new_group_id[group_vec[nid]] != -1) {
+        // propagate new group id from child
+        group_vec[nid] = new_group_id[group_vec[nid]];
+      }
+      TOpPattern pt = op_pattern.get(idx[nid].source->op(), kOpaque);
+      if (pt == kOpaque) continue;
+      const auto& group_ids = children_group_ids[nid];
+      if (group_ids.size() <= 1) continue;
+      const auto child_group_id = group_ids[0];
+      // fuse this node with children if all children belong to the same group
```

Comment:
We need a stronger rule, currently, simply merges all of the children may not be good for future operations

Reply:
I can add a check to see if all children are elem wise ops.

But the only case I fuse a parent node with children is when the parent has multiple children, and all children nodes are already fused by the existing algorithm. I think this only occurs when all children are elem wise ops.

Do you have a counter example that fails my logic?

Reply:
I think it is fine for now, but we still want to be careful and explicit here, in case there are future updates to the code that brings the rule